// @replayio/playwright/dist/_bundled/@replay-cli/shared/recording/types.js
'use strict';

var RECORDING_LOG_KIND = /* @__PURE__ */ ((RECORDING_LOG_KIND2) => {
  RECORDING_LOG_KIND2["addMetadata"] = "addMetadata";
  RECORDING_LOG_KIND2["crashData"] = "crashData";
  RECORDING_LOG_KIND2["crashed"] = "crashed";
  RECORDING_LOG_KIND2["crashUploaded"] = "crashUploaded";
  RECORDING_LOG_KIND2["createRecording"] = "createRecording";
  RECORDING_LOG_KIND2["originalSourceAdded"] = "originalSourceAdded";
  RECORDING_LOG_KIND2["processingFailed"] = "processingFailed";
  RECORDING_LOG_KIND2["processingFinished"] = "processingFinished";
  RECORDING_LOG_KIND2["processingStarted"] = "processingStarted";
  RECORDING_LOG_KIND2["recordingUnusable"] = "recordingUnusable";
  RECORDING_LOG_KIND2["sourcemapAdded"] = "sourcemapAdded";
  RECORDING_LOG_KIND2["uploadFailed"] = "uploadFailed";
  RECORDING_LOG_KIND2["uploadFinished"] = "uploadFinished";
  RECORDING_LOG_KIND2["uploadStarted"] = "uploadStarted";
  RECORDING_LOG_KIND2["writeFinished"] = "writeFinished";
  RECORDING_LOG_KIND2["writeStarted"] = "writeStarted";
  return RECORDING_LOG_KIND2;
})(RECORDING_LOG_KIND || {});

exports.RECORDING_LOG_KIND = RECORDING_LOG_KIND;
